README

INTRODUCTION
------------

The vblade is the virtual EtherDrive (R) blade, a program that makes a
seekable file available over an ethernet local area network (LAN) via
the ATA over Ethernet (AoE) protocol.

The seekable file is typically a block device like /dev/md0 but even
regular files will work.  Sparse files can be especially convenient.
When vblade exports the block storage over AoE it becomes a storage
target.  Another host on the same LAN can access the storage if it has
a compatible aoe kernel driver.

BUILDING
--------

The following command should build the vblade program on a Linux-based
system:

  make

For FreeBSD systems, include an extra parameter like so:

  make PLATFORM=freebsd

EXAMPLES
--------

There is a "vbladed" script that daemonizes the program and sends its
output to the logger program.  Make sure you have logger installed if
you would like to run vblade as a daemon with the vbladed script.

  ecashin@kokone vblade$ echo 'I have logger' | logger
  ecashin@kokone vblade$ tail -3 /var/log/messages
  Feb  8 14:52:49 kokone -- MARK --
  Feb  8 15:12:49 kokone -- MARK --
  Feb  8 15:19:56 kokone logger: I have logger

Here is a short example showing how to export a block device with a
vblade.  (This is a loop device backed by a sparse file, but you could
use any seekable file instead of /dev/loop7.)

  ecashin@kokone vblade$ make
  cc -Wall   -c -o aoe.o aoe.c
  cc -Wall   -c -o linux.o linux.c
  cc -Wall   -c -o ata.o ata.c
  cc -o vblade aoe.o linux.o ata.o
  ecashin@kokone vblade$ su
  Password: 
  root@kokone vblade# modprobe loop
  root@kokone vblade# dd if=/dev/zero bs=1k count=1 seek=`expr 1024 \* 4096` of=bd
  -file
  1+0 records in
  1+0 records out
  1024 bytes transferred in 0.009901 seconds (103423 bytes/sec)
  root@kokone vblade# losetup /dev/loop7 bd-file  
  root@kokone vblade# ./vblade 9 0 eth0 /dev/loop7 
  ioctl returned 0
  4294968320 bytes
  pid 16967: e9.0, 8388610 sectors

Here's how you can use the Linux aoe driver to access the storage from
another host on the LAN.

  ecashin@kokone ecashin$ ssh makki
  Last login: Mon Feb  7 10:25:04 2005
  ecashin@makki ~$ su
  Password: 
  root@makki ecashin# modprobe aoe
  root@makki ecashin# aoe-stat
      e9.0            eth1              up
  root@makki ecashin# mkfs -t ext3 /dev/etherd/e9.0
  mke2fs 1.35 (28-Feb-2004)
...
  Creating journal (8192 blocks): done
  Writing superblocks and filesystem accounting information: done
  
  This filesystem will be automatically checked every 24 mounts or
  180 days, whichever comes first.  Use tune2fs -c or -i to override.
  root@makki ecashin# mkdir /mnt/e9.0
  root@makki ecashin# mount /dev/etherd/e9.0 /mnt/e9.0
  root@makki ecashin# echo hooray > /mnt/e9.0/test.txt
  root@makki ecashin# cat /mnt/e9.0/test.txt
  hooray

Remember: be as careful with these devices as you would with /dev/hda!

Jumbo Frame Compatibility
-------------------------

Vblade can use jumbo frames provided your initiator is jumbo frame
capable.  There is one small configuration gotcha to consider 
to avoid having the vblade kernel frequently drop frames.

Vblade uses a raw socket to perform AoE.  The linux kernel will
only buffer a certain amount of data for a raw socket.  For 2.6
kernels, this value is managed through /proc:

root@nai aoe# grep . /proc/sys/net/core/rmem_*
/proc/sys/net/core/rmem_default:128000
/proc/sys/net/core/rmem_max:128000

rmem_max is the max amount a user process may expand the receive
buffer to -- through setsockopt(...) -- and rmem_default is, as you
might expect, the default.

The gotcha is that this amount to buffer does not relate
to the amount of user data buffered, but the amount of
real data buffered.  As an example, the Intel GbE controller
must be given 16KB frames to use an MTU over 8KB.
For each received frame, the kernel must be able to buffer
16KB, even if the aoe frame is only 60 bytes in length.

The linux aoe initiator will use 16 outstanding frames when
used with vblade.  A good default for ensuring frames are
not dropped is to allocate 16KB for 17 frames:

for f in /proc/sys/net/core/rmem_*; do echo $((17 * 16 * 1024)) >$f; done

Be sure to start vblade after changing the buffering defaults
as the buffer value is set when the socket is opened.

AoE Initiator Compatibility
---------------------------

The Linux aoe driver for the 2.6 kernel is compatible if you use
aoe-2.6-7 or newer.  You can use older aoe drivers but you will only
be able to see one vblade per MAC address.

Contrib Patches
---------------

see contrib/README

Kvblade
-------

  While vblade runs as a userland process (like "ls" or "vi"), there
  is another program that runs inside the kernel.  It is called
  kvblade.  It is alpha software.


NEWS
-*- change-log -*-
2020-12-02 Andreas Kölbl <koelblandreas@freenet.de>
	Support recent compilers by avoiding multiple definitions
	vblade-25

2018-08-25 Christoph Biedl <sourceforge.bnwi@manchmal.in-ulm.de>
	Print helpful error and exit immediately for missing device
	vblade-24

2017-11-19 Catalin Salgau <csalgau@users.sourceforge.net>
	On FreeBSD limit used MTU to address BPF limitation

2015-06-14 Ed Cashin <ed.cashin@acm.org>
	Add convenience script for creating sparse files
	vblade-23

2015-02-25 Catalin Salgau <csalgau@users.sourceforge.net>
	Warn about Windows problem with CHS misalignment

2014-09-13 Ed Cashin <ed.cashin@acm.org>
	code cleanup: remove unused variables

2014-08-10 Ed Cashin <ed.cashin@acm.org>
	vblade-22

2014-06-08 Ed Cashin <ed.cashin@acm.org>
	update version for v22 release candidate 1
	buffer boundary cleanups
	FreeBSD BPF and MTU fixes from Catalin Salgau
	offset and size options by Christoph Biedl
	vblade-22rc1

2013-03-18 Ed Cashin <ecashin@coraid.com>
	add big-endian support from Daniel Mealha Cabrita <dancab@gmx.net>
	vblade-21

2009-08-14 Sam Hopkins <sah@coraid.com>
	bugfix: aoe command error did not set Error bit in flags
	add support for AoEr11
	set ident serial to shelf.slot:hostname
	vblade-20

2008-10-08 Ed Cashin <ecashin@coraid.com>
	add Chris Webb's bpf fix for FreeBSD
	add Ryan Thomas's fix to stop bufcnt being overridden
	vblade-19
	
2008-07-14 Ed Cashin <ecashin@coraid.com>
	add Chris Webb's block device options patch
	add Chris Webb's socket options patch for better jumbo handling
	remove obsolete contrib/o_direct.diff
	vblade-18

2008-06-09 Ed Cashin <ecashin@coraid.com>
	add Chris Webb's latest BPF patch to vblade, remove from contrib
	update contributed AIO patch for compatibility with current sources
	vblade-17
	
2008-05-07 Ed Cashin <ecashin@coraid.com>
	add Chris Webb's AIO patch to the contributions
	add Chris Webb's BPF patch to the contributions
	vblade-16

2008-02-20 Ed Cashin <ecashin@coraid.com>
	require the amount of data we use, not the amount ethernet requires
	make sure the packet length agrees with the config query length
	make sure the packet length agrees with the amount to write
	remove newline embedded in fw version field of ATA dev ID response
	vblade-15
	
2006-11-20 Sam Hopkins <sah@coraid.com>
	apply contrib jumbo patch to standard distribution
	add jumbo configuration app. note in README
	add jumbo README reference to manpage
	add mask feature; -m flag
	update manpage to describe -m flag
	vblade-14

2006-10-05 Sam Hopkins <sah@coraid.com>
	fix confcmd memcpy bug
	correct scnt return value in read/write ata response
	replace O_RDONLY fallback with explicit stat.  root always wins.
	vblade-13

2006-10-04 Sam Hopkins <sah@coraid.com>
	fix confcmd buglets
	fix atacmd buglets
	add atacmd handling for bad argument errors
	add O_RDONLY open if O_RDWR fails
	add contrib patch directory
	add contrib/README
	add jumbo patch to contrib
	add o_direct patch to contrib
	vblade-12

2006-09-21 "Adam J. Richter" <adam@yggdrasil.com>
	add install target for makefile
	vblade-11
	
2005-12-06 Ed Cashin <ecashin@coraid.com>
	fix u64 configuration on FreeBSD
	release vblade-10
	
2005-12-06 Valeriy Glushkov <valery@rocketdivision.com>
	implemented config string support
	added handler for ATA Check power mode command
	
2005-11-15 Ed Cashin <ecashin@coraid.com>
	add compatibility with platforms lacking u64 (e.g., Slackware)
	release vblade-9
	
2005-11-10 Ed Cashin <ecashin@coraid.com>
	call atainit on program startup
	put VBLADE_VERSION in dat.h and use it in firmware version
	release vblade-7
	include Stacey's patch to use p{read,write} on FreeBSD
	include Stacey's patch to typedef ulong on FreeBSD
	fix makefile dependencies (e.g., rebuild on new aoe.c)
	fix config string length specification
	include Stacey's patch to avoid compile warnings on FreeBSD
	release vblade-8
	
2005-11-10 "Stacey D. Son" <sson@verio.net>
	include FreeBSD support
	
2005-10-03 Ed Cashin <ecashin@coraid.com>
	don't invoke vblade with dash from vbladed
	
2005-08-31 20:14:12 GMT Ed Cashin <ecashin@coraid.com>
	ATA identify: don't juggle bytes in shorts on big endian arch
	add manpage for vblade, vbladed
	release vblade-6
	
2005-03-17 15:24:30 GMT	Ed Cashin <ecashin@coraid.com>
	follow up on vblade-2's off-by-one patch, making end of device usable
	release vblade-5
	
2005-03-15 22:03:17 GMT	Ed Cashin <ecashin@coraid.com>
	don't rely on kernel headers for defining the aoe type 0x88a2
	release vblade-4
	
2005-03-15 17:27:01 GMT	Ed Cashin <ecashin@coraid.com>
	docs: aoe-2.6-7 is the first driver to support multiple blades per mac
	release vblade-3
	
2005-03-11 18:30:26 GMT	Ed Cashin <ecashin@coraid.com>
	put 64-bit configuration into config.h file
	don't use uninitialized variables
	broadcast config query on startup
	clarify desired patch format in HACKING
	add sah@coraid.com's vblade-1.ata.c.patch: fix off-by-one and ext LBA
	add docs, remove daemonizing code from vblade
	release vblade-2
	
2005-02-08 20:21:52 GMT	Ed Cashin <ecashin@coraid.com>
	starting documentation
	add script that daemonizes vblade process, logging output
	make vblade sources -Wall clean, use daemon(3)
	release vblade-1


ata.c
// ata.c:  ATA simulator for vblade
#include "config.h"
#include <string.h>
#include <stdio.h>
#include <sys/types.h>
#include "dat.h"
#include "fns.h"

enum {
	// err bits
	UNC =	1<<6,
	MC =	1<<5,
	IDNF =	1<<4,
	MCR =	1<<3,
	ABRT = 	1<<2,
	NM =	1<<1,

	// status bits
	BSY =	1<<7,
	DRDY =	1<<6,
	DF =	1<<5,
	DRQ =	1<<3,
	ERR =	1<<0,
};

static ushort ident[256];

static void
setfld(ushort *a, int idx, int len, char *str)	// set field in ident
{
	uchar *p;

	p = (uchar *)(a+idx);
	while (len > 0) {
		if (*str == 0)
			p[1] = ' ';
		else
			p[1] = *str++;
		if (*str == 0)
			p[0] = ' ';
		else
			p[0] = *str++;
		p += 2;
		len -= 2;
	}
}

static void
setlba28(ushort *ident, vlong lba)
{
	uchar *cp;

	cp = (uchar *) &ident[60];
	*cp++ = lba;
	*cp++ = lba >>= 8;
	*cp++ = lba >>= 8;
	*cp++ = (lba >>= 8) & 0xf;
}

static void
setlba48(ushort *ident, vlong lba)
{
	uchar *cp;

	cp = (uchar *) &ident[100];
	*cp++ = lba;
	*cp++ = lba >>= 8;
	*cp++ = lba >>= 8;
	*cp++ = lba >>= 8;
	*cp++ = lba >>= 8;
	*cp++ = lba >>= 8;
}

static void
setushort(ushort *a, int i, ushort n)
{
	uchar *p;

	p = (uchar *)(a+i);
	*p++ = n & 0xff;
	*p++ = n >> 8;
}

void
atainit(void)
{
	char buf[64];

	setushort(ident, 47, 0x8000);
	setushort(ident, 49, 0x0200);
	setushort(ident, 50, 0x4000);
	setushort(ident, 83, 0x5400);
	setushort(ident, 84, 0x4000);
	setushort(ident, 86, 0x1400);
	setushort(ident, 87, 0x4000);
	setushort(ident, 93, 0x400b);
	setfld(ident, 27, 40, "Coraid EtherDrive vblade");
	sprintf(buf, "V%d", VBLADE_VERSION);
	setfld(ident, 23, 8, buf);
	setfld(ident, 10, 20, serial);
}


/* The ATA spec is weird in that you specify the device size as number
 * of sectors and then address the sectors with an offset.  That means
 * with LBA 28 you shouldn't see an LBA of all ones.  Still, we don't
 * check for that.
 */
int
atacmd(Ataregs *p, uchar *dp, int ndp, int payload) // do the ata cmd
{
	vlong lba;
	ushort *ip;
	int n;
	enum { MAXLBA28SIZE = 0x0fffffff };
	extern int maxscnt;

	p->status = 0;
	switch (p->cmd) {
	default:
		p->status = DRDY | ERR;
		p->err = ABRT;
		return 0;
	case 0xe7:		// flush cache
		return 0;
	case 0xec:		// identify device
		if (p->sectors != 1 || ndp < 512)
			return -1;
		memmove(dp, ident, 512);
		ip = (ushort *)dp;
		if (size & ~MAXLBA28SIZE)
			setlba28(ip, MAXLBA28SIZE);
		else
			setlba28(ip, size);
		setlba48(ip, size);
		p->err = 0;
		p->status = DRDY;
		p->sectors = 0;
		return 0;
	case 0xe5:		// check power mode
		p->err = 0;
		p->sectors = 0xff; // the device is active or idle
		p->status = DRDY;
		return 0;
	case 0x20:		// read sectors
	case 0x30:		// write sectors
		lba = p->lba & MAXLBA28SIZE;
		break;
	case 0x24:		// read sectors ext
	case 0x34:		// write sectors ext
		lba = p->lba & 0x0000ffffffffffffLL;	// full 48
		break;
	}

	// we ought not be here unless we are a read/write

	if (p->sectors > maxscnt || p->sectors*512 > ndp)
		return -1;

	if (lba + p->sectors > size) {
		p->err = IDNF;
		p->status = DRDY | ERR;
		p->lba = lba;
		return 0;
	}
	if (p->cmd == 0x20 || p->cmd == 0x24)
		n = getsec(bfd, dp, lba, p->sectors);
	else {
		// packet should be big enough to contain the data
		if (payload < 512 * p->sectors)
			return -1;
		n = putsec(bfd, dp, lba, p->sectors);
	}
	n /= 512;
	if (n != p->sectors) {
		p->err = ABRT;
		p->status = ERR;
	} else
		p->err = 0;
	p->status |= DRDY;
	p->lba += n;
	p->sectors -= n;
	return 0;
}

